File: NatureOfCodeCocos2DX/Classes/Aurora/Random/RandomBase.cpp
//
//  RandomBase.cpp
//  NatureOfCodeCocos2DX
//
//  Created by Adrian Simionescu on 08/10/14.
//
//

#include "RandomBase.h"


namespace Aurora {
    namespace Random {
		//-------------------------------------------------------------------------
		// START = Random Base Interfaces
		//-------------------------------------------------------------------------
		PRandomBase::PRandomBase()
		{

		}
		
		PRandomBase::~PRandomBase()
		{

		}

		PRandomBaseBasic::PRandomBaseBasic() : PRandomBase()
		{

		}

		PRandomBaseBasic::~PRandomBaseBasic()
		{
			PRandomBase::~PRandomBase();
		}

		PRandomBaseNormal::PRandomBaseNormal() : PRandomBaseBasic()
		{

		}

		PRandomBaseNormal::~PRandomBaseNormal()
		{
			PRandomBaseBasic::~PRandomBaseBasic();
		}

		PRandomBaseUniform::PRandomBaseUniform() : PRandomBaseBasic()
		{

		}

		PRandomBaseUniform::~PRandomBaseUniform()
		{
			PRandomBaseBasic::~PRandomBaseBasic();
		}

		PRandomBaseGaussian::PRandomBaseGaussian() : PRandomBaseBasic()
		{

		}

		PRandomBaseGaussian::~PRandomBaseGaussian()
		{
			PRandomBaseBasic::~PRandomBaseBasic();
		}

		PRandomBasePerlinNoise::PRandomBasePerlinNoise() : PRandomBaseBasic()
		{

		}

		PRandomBasePerlinNoise::~PRandomBasePerlinNoise()
		{
			PRandomBaseBasic::~PRandomBaseBasic();
		}

		PRandomBaseComplete::PRandomBaseComplete() : PRandomBaseUniform(), PRandomBaseGaussian(), PRandomBasePerlinNoise()
		{
			
		}

		PRandomBaseComplete::~PRandomBaseComplete()
		{
			PRandomBaseUniform::~PRandomBaseUniform();
			PRandomBaseGaussian::~PRandomBaseGaussian();
			PRandomBasePerlinNoise::~PRandomBasePerlinNoise();
		}

		//-------------------------------------------------------------------------
		// END = Random Base Interfaces
		//-----------------------------------------------------------------------

        RandomBase::RandomBase() : PRandomBaseBasic()
        {
            this->RandomNumberModeValue = RandomNumberMode::Normal;
        }

		RandomBase::~RandomBase()
		{
			PRandomBaseBasic::~PRandomBaseBasic();
		}
        
        void RandomBase::SetConstrainsRange(Float width, Float height)
        {
			this->Width = width;
			this->Height = height;
        }
        
        void RandomBase::Constrain()
        {
			
        }
        
        void RandomBase::DoCalculations()
        {
			this->Constrain();
        }
        
        void RandomBase::SetRandomNumberMode(Aurora::Random::RandomNumberMode randomNumberMode)
        {
			this->RandomNumberModeValue = randomNumberMode;
        }

		

		RandomBaseComplete::RandomBaseComplete() : RandomBase()
		{

		}

		RandomBaseComplete::~RandomBaseComplete()
		{
			RandomBase::~RandomBase();
			PRandomBaseComplete::~PRandomBaseComplete();
		}

		void RandomBaseComplete::DoCalculations()
		{
			RandomBase::DoCalculations();

			switch(this->RandomNumberModeValue)
            {
			case RandomNumberMode::Normal:
				{
					this->NormalCalculations();
				} break;

			case RandomNumberMode::Uniform:
				{
					this->UniformCalculations();
				} break;


			case RandomNumberMode::Gaussian:
				{
					this->GaussianCalculations();
				} break;

			case RandomNumberMode::Perlin:
				{
					this->PerlinNoiseCalculations();
				} break;

			default:
				{
					this->NormalCalculations();
				} break;
			}
  
        
			this->Constrain();
		}

		void RandomBaseComplete::Constrain()
		{
			RandomBase::Constrain();

		}

		void RandomBaseComplete::NormalCalculations()
		{
			
		}

		void RandomBaseComplete::UniformCalculations()
		{

		}

		void RandomBaseComplete::GaussianCalculations()
		{

		}

		void RandomBaseComplete::PerlinNoiseCalculations()
		{

		}

		RandomBasePerlinNoise::RandomBasePerlinNoise() : RandomBase()
		{

		}

		RandomBasePerlinNoise::~RandomBasePerlinNoise()
		{
			RandomBase::~RandomBase();
			PRandomBasePerlinNoise::~PRandomBasePerlinNoise();
		}

		void RandomBasePerlinNoise::DoCalculations()
		{
			RandomBase::DoCalculations();
			this->PerlinNoiseCalculations();
		}

		void RandomBasePerlinNoise::PerlinNoiseCalculations()
		{

		}

		RandomBaseGaussian::RandomBaseGaussian() : RandomBase()
		{

		}

		RandomBaseGaussian::~RandomBaseGaussian()
		{
			RandomBase::~RandomBase();
			PRandomBaseGaussian::~PRandomBaseGaussian();
		}

		void RandomBaseGaussian::DoCalculations()
		{
			RandomBase::DoCalculations();
			this->GaussianCalculations();
		}

		void RandomBaseGaussian::GaussianCalculations()
		{

		}

		RandomBaseUniform::RandomBaseUniform() : RandomBase()
		{

		}

		RandomBaseUniform::~RandomBaseUniform()
		{
			RandomBase::~RandomBase();
			PRandomBaseUniform::~PRandomBaseUniform();
		}

		void RandomBaseUniform::DoCalculations()
		{
			RandomBase::DoCalculations();
			this->UniformCalculations();
		}

		void RandomBaseUniform::UniformCalculations()
		{

		}

		RandomBaseNormal::RandomBaseNormal() : RandomBase()
		{

		}

		RandomBaseNormal::~RandomBaseNormal()
		{
			RandomBase::~RandomBase();
			PRandomBaseNormal::~PRandomBaseNormal();
		}

		void RandomBaseNormal::DoCalculations()
		{
			RandomBase::DoCalculations();
			this->NormalCalculations();
		}

		void RandomBaseNormal::NormalCalculations()
		{

		}
        
    }; // END OF NAMESPACE Random
}; // END OF NAMESPACE Aurora

File: NatureOfCode in Swift(Retired)/NatureOfCode/RandomBase.swift
//
//  RandomBase.swift
//  NatureOfCode
//
//  Created by Adrian Simionescu on 30/09/14.
//  Copyright (c) 2014 Adrian Simionescu. All rights reserved.
//

import Foundation

public enum RandomNumberMode
{
    case Uniform, Gaussian, Perlin
}

protocol PRandomBase
{
    var RandomNumberModeValue : RandomNumberMode { get set };
    
    func SetRandomNumberMode( randomNumberMode : RandomNumberMode);
}

protocol PRandomBaseBasic : PRandomBase
{
    var Width : CGFloat { get set };
    var Height : CGFloat { get set };
    
    func SetConstrainsRange( width : CGFloat, height: CGFloat);
    func DoCalculations();
    func Constrain();
    func Normal();
}

protocol PRandomBaseUniform : PRandomBaseBasic
{
    mutating func UniformCalculations();
}

protocol PRandomBaseGaussian : PRandomBaseBasic
{
    mutating func GaussianCalculations();
}

protocol PRandomBasePerlinNoise : PRandomBaseBasic
{
    var PerlinNoiseCalculator : PerlinNoise { get set };
    
    mutating func PerlinNoiseCalculations();
}

protocol PRandomBaseComplete : PRandomBaseUniform, PRandomBaseGaussian, PRandomBasePerlinNoise
{
    
}

public class RandomBase : PRandomBaseBasic
{
    var RandomNumberModeValue : RandomNumberMode;
    var Width : CGFloat = 0.0;
    var Height : CGFloat = 0.0;
    
    init()
    {
        RandomNumberModeValue = RandomNumberMode.Uniform;
    }
    
    func SetRandomNumberMode( randomNumberMode : RandomNumberMode)
    {
        self.RandomNumberModeValue = randomNumberMode;
    }
    
    func SetConstrainsRange( width : CGFloat, height: CGFloat)
    {
        self.Width = width;
        self.Height = height;
    }
    
    func Constrain()
    {
        
    }
    
    func DoCalculations() {
        self.Constrain();
    }

    func Normal() {

    }
}

public class RandomBaseComplete : RandomBase, PRandomBaseComplete
{
    var PerlinNoiseCalculator : PerlinNoise = PerlinNoise(seed: 10);
    
    override func DoCalculations()
    {
        super.DoCalculations();
        switch(self.RandomNumberModeValue)
            {
        case .Uniform:
            self.UniformCalculations();
            
        case .Gaussian:
            self.GaussianCalculations();
        case .Perlin:
            self.PerlinNoiseCalculations();

        default:
            self.UniformCalculations();
        }
        
        self.Constrain();
    }
    
    override func Constrain()
    {
        super.Constrain();
    }
    
    override func Normal()
    {
        
    }
    
    func UniformCalculations()
    {
        
    }
    
    func GaussianCalculations()
    {
        
    }
    
    func PerlinNoiseCalculations()
    {
    }
}

public class RandomBasePerlinNoise : RandomBase, PRandomBasePerlinNoise
{
    var PerlinNoiseCalculator : PerlinNoise = PerlinNoise(seed: 10);
    
    override func DoCalculations()
    {
        super.DoCalculations();
        self.PerlinNoiseCalculations();
        
    }
    
    func PerlinNoiseCalculations()
    {
        
    }
}

public class RandomBaseGaussian : RandomBase, PRandomBaseGaussian
{
    override func DoCalculations()
    {
        super.DoCalculations();
        self.GaussianCalculations();
        
    }
    
    func GaussianCalculations()
    {
        
    }
}

public class RandomBaseUniform : RandomBase, PRandomBaseUniform
{
    override func DoCalculations()
    {
        super.DoCalculations();
        self.UniformCalculations();
        
    }
    
    func UniformCalculations()
    {
        
    }
}

